File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int n = s.size();
        unordered_map<char, int> frequencyMap;
        int res = 0;
        int l = 0;

        for (int r = 0; r < n; ++r) {
            frequencyMap[s[r]]++;
            while (l < r && frequencyMap['a'] > 0 && frequencyMap['b'] > 0 && frequencyMap['c'] > 0) {
                res += n - r;
                frequencyMap[s[l]]--;
                l++;
            }
        }
        return res;
    }
};

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        frequencyMap = defaultdict(int)
        res = 0
        l = 0

        for r in range(n):
            frequencyMap[s[r]] += 1
            while l < r and all(frequencyMap[c] > 0 for c in ['a','b', 'c']):
                res += n-r
                frequencyMap[s[l]] -= 1
                l += 1
        return res
            